src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt

```diff
@@ -520,17 +554,56 @@ private fun RsCallExpr.expectedParamsCount(): Int? {
     }
 }
 
+private fun RsMethodCallExpr.expectedMutableSelfParameter(): Boolean {
+    val fn = reference.resolve() as? RsFunction ?: return false
+    val self = fn.selfParameter ?: return false
+    return self.isMut && self.isRef
+}
+
 private fun RsMethodCallExpr.expectedParamsCount(): Int? {
     val fn = reference.resolve() as? RsFunction ?: return null
     if (fn.queryAttributes.hasCfgAttr()) return null
     return if (fn.isInherentImpl) fn.valueParameterList?.valueParameterList?.size else null
 }
 
+private val RsCallExpr.declaration: RsFunction?
+    get() = (expr as? RsPathExpr)?.path?.reference?.resolve() as? RsFunction
+
+private fun RsExpr.declaration(): RsCompositeElement? = when (this) {
+    is RsPathExpr -> this.path.reference.resolve()
+    is RsFieldExpr -> this.expr.declaration()
+    else -> null
+}
+
 private fun RsExpr.isMutable(): Boolean {
-    val path = (this as? RsPathExpr)?.path ?: return true
-    val declaration = path.reference.resolve()
+    return when (this) {
+        is RsPathExpr -> {
+            val declaration = this.path.reference.resolve() ?: return true
+
+            if (declaration is RsSelfParameter) {
+                return declaration.mut != null
+            }
+
+            if (declaration is RsPatBinding && declaration.isMut) {
+                return true
+            }
 
-    declaration?.parentOfType<RsLetDecl>()?.eq ?: return true
+            val type = this.type
+            if (type is RustReferenceType) {
+                return type.mutable
+            }
 
-    return (declaration as? RsPatBinding)?.isMut ?: true
+            val letExpr = declaration.parentOfType<RsLetDecl>()
+            if (letExpr != null && letExpr.eq == null) {
+                return true
+            }
+            if (type is RustUnknownType) {
+                return true
+            }
+            false
+        }
+        is RsFieldExpr -> (this.expr.type as? RustReferenceType)?.mutable ?: true
+        is RsUnaryExpr -> mul != null || (expr != null && expr?.isMutable()!!)
```

Comment:
Are these `!!` necessary to compile? If so, then you can stick with older version `(expr?.isMutable() ?: true)`